nixos/modules/services/web-apps/librechat.nix

```
@@ -0,0 +1,162 @@
+      default = "librechat";
+      example = "users";
+      description = "The group to run the service as.";
+    };
+    credentials = lib.mkOption {
+      type =
+        with lib.types;
+        submodule {
+          freeformType = attrsOf path;
+        };
+      default = { };
+      example = {
+        CREDS_KEY = /run/secrets/creds_key;
+      };
+      description = "Environment variables which are loaded from the contents of files at a file paths, mainly used for secrets. See https://www.librechat.ai/docs/configuration/dotenv for a full list.";
+    };
+    env = lib.mkOption {
+      type =
+        with lib.types;
+        submodule {
+          freeformType = attrsOf (oneOf [
+            str
+            path
+            int
+            float
+          ]);
+        };
+      example = {
+        ALLOW_REGISTRATION = "true";
+        HOST = "0.0.0.0";
+        CONSOLE_JSON_STRING_LENGTH = 255;
+      };
+      default = { };
+      description = "Environment variables that will be set for the service. See https://www.librechat.ai/docs/configuration/dotenv for a full list.";
+    };
+    settings = lib.mkOption {
+      type = lib.types.submodule {
+        freeformType = format.type;
+      };
+      default = { };
+      example = # nix
+        ''
+          {
+            version = "1.0.8";
+            cache = true;
+            interface = {
+              privacyPolicy = {
+                externalUrl = "https://librechat.ai/privacy-policy";
+                openNewTab = true;
+              };
+            };
+            endpoints = {
+              custom = [
+                {
+                  name = "OpenRouter";
+                  # Note that the following $ should be escaped with a backslash, not '''
+                  apiKey = "''${OPENROUTER_KEY}";
+                  baseURL = "https://openrouter.ai/api/v1";
+                  models = {
+                    default = ["meta-llama/llama-3-70b-instruct"];
+                    fetch = true;
+                  };
+                  titleConvo = true;
+                  titleModule = "meta-llama/llama-3-70b-instruct";
+                  dropParams = ["stop"];
+                  modelDisplayLabel = "OpenRouter";
+                }
+              ];
+            };
+          }
+        '';
+      description = "A free-form attribute set that will be written to librechat.yaml. You can use environment variables by wrapping them in \${}. Take care to escape the \$ character.";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      (lib.mkIf (cfg.env ? PORT) {
+        assertion = cfg.port == cfg.env.PORT;
+        message = "You cannot set services.librechat.port and services.librechat.env.PORT to different values.";
+      })
+    ];
+    networking.firewall.allowedTCPPorts = lib.optionals cfg.openFirewall [ cfg.port ];
+    systemd.tmpfiles.rules = [
+      "d ${cfg.logDir} ${"0755"} ${cfg.user} ${cfg.group} -"
+    ];
```

It's just my personal opinion but IMO the systemd.tmpfiles.settings API looks better

```suggestion
    systemd.tmpfiles.settings."10-librechat"."${cfg.logDir}".d = {
      mode = "0755";
      inherit (cfg) user group;
    ];
```